File: QTVI-refactor/4_find_wave_bounds/stdoutlier.cpp
// ============================================================================
// File: stdoutlier.cpp
// ============================================================================
#include "stdoutlier.h"
#include "StatsUtils.h"

vector<bool> stdoutlier(const vector<double>& data,
    double multiplier,
    size_t mean_window,
    const string& direction,
    bool debug_plot) {

    vector<double> d = diff(data);
    vector<double> x = movmean(d, mean_window);
    double s = std_dev(d);

    vector<double> upper_bound(d.size());
    vector<double> lower_bound(d.size());
    for (size_t i = 0; i < d.size(); ++i) {
        upper_bound[i] = x[i] + s * multiplier;
        lower_bound[i] = x[i] - s * multiplier;
    }

    vector<bool> outlierDiffs(d.size(), false);
    if (direction == "lower") {
        for (size_t i = 0; i < d.size(); ++i)
            outlierDiffs[i] = (d[i] < lower_bound[i]);
    }
    else if (direction == "upper") {
        for (size_t i = 0; i < d.size(); ++i)
            outlierDiffs[i] = (d[i] > upper_bound[i]);
    }
    else {
        for (size_t i = 0; i < d.size(); ++i)
            outlierDiffs[i] = (d[i] > upper_bound[i] || d[i] < lower_bound[i]);
    }

    // Mark both endpoints of each outlier diff
    vector<bool> outliers(data.size(), false);
    for (size_t i = 0; i + 1 < data.size(); ++i) {
        if (i < outlierDiffs.size() && outlierDiffs[i]) {
            outliers[i] = true;
            outliers[i + 1] = true;
        }
    }

    return outliers;
}

File: QTVI-Daniel/helpers/general/stdoutlier.m
function [outliers] = stdoutlier(data, multiplier, mean_window, direction, debug_plot)
    %% data == nx1 or 1xn matrix, Multiplier == std multipler (normally ~2.5 or 3)
    %% mean_window == length(data) * .02 normally good value
    x = movmean(diff(data), mean_window);
    s = std(diff(data));
    d = diff(data);
    upper_bound = x + s * multiplier;
    lower_bound = x - s * multiplier;
    if strcmp(direction,'lower')
        weridOnes = (d < lower_bound);
    elseif strcmp(direction,'upper')
        weridOnes = (d > upper_bound);
    else
        weridOnes = (d > upper_bound | d < lower_bound);
    end
    
    if debug_plot
        figure;
        plot(d, '.');
        hold on;
        plot(x);
        plot(upper_bound);
        plot(lower_bound);
        t = (1:length(d));
        plot(t(weridOnes == 1), d(weridOnes == 1), 'or');
        hold off;
    end

    outliers = zeros(length(data), 1);

    for i = 1:length(data) - 1

        if weridOnes(i) == 1
            outliers(i:i + 1) = 1;
        end

    end

    outliers = logical(outliers);
end